Add PAM algorithm to K-Medoids

TimotheeMathieu opened the pull request:
Following issue #46 here is an implementation of PAM algorithm.

* I only implemented the swap part of the algorithm as we do not need the greedy initialization (build part) because we already have initialization algorithms available.
* PAM algorithm is rather slow, at least it is slower than the method used until now.
* I did not find the name of the method used until now. For now, I named it "fast", if some of you have a better idea, tell me.
* When used for `plot_kmedoids_digits.py`, PAM gives the same results as "fast" method, it is just slower.

PROBLEM:
PAM is very slow: on `plot_kmedoids_digits.py` PAM runs in 2min30s against 3s for the "fast" implementation on my computer. Maybe we should implement the algorithm FastPAM from https://arxiv.org/abs/1810.05691 which is used for instance in R (https://www.rdocumentation.org/packages/cluster/versions/2.1.0/topics/pam see argument `pamonce`)

rth commented:
Thanks for looking into this @TimotheeMathieu !

> PAM algorithm is rather slow, at least it is slower than the method used until now.

Indeed, it's around 1000x times slower than the current method on the example,
```
        wall_time (s)                                                                                                                      
method           
pam     16.112913
fast     0.025839
```
As far as I understand it's expected that PAM is slower, but I think the current implementation might also be improved performance wise (more on that in the following comment).  

>  For now, I named it "fast", if some of you have a better idea, tell me.

Since there is FastPAM paper, we would probably need a more explicit name. Can't think of anything right now.

We would need to add tests that check that the output is the same between methods, and also ideally check that the inertia is smaller for the problematic cases mentionned in https://github.com/scikit-learn-contrib/scikit-learn-extra/pull/44#issuecomment-569142015 

Maybe @kno10 would have some suggestions how to improve this implementation, or might even be willing to contribute the FastPAM implementation (in a separate PR) :)

rth commented:
I run a quick profiling of the example,

<details>

```py
from sklearn_extra.cluster import KMedoids
from sklearn.datasets import load_digits
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale


digits = load_digits()
data = scale(digits.data)
reduced_data = PCA(n_components=10).fit_transform(data)
n_digits = len(np.unique(digits.target))


KMedoids(metric="manhattan", n_clusters=n_digits, method="pam").fit(reduced_data)
```
</details>

The output is below,
![image](https://user-images.githubusercontent.com/630936/98447395-cdc96800-2124-11eb-844e-c16676415783.png)

![image](https://user-images.githubusercontent.com/630936/98448396-01f45700-212c-11eb-89f2-e164edcb8b21.png)


So most of the time is indeed spent computing the cost function and in particular the min function which we call 600k times.

I think re-implementing the critical parts in Cython would probably help significantly.

@jeremiedbb might also have some insights. Maybe there are Cython functions in scikit-learn we could reuse.

kno10 commented:
The original PAM does not use `compute_inertia`.

Computing this naively supposedly adds another factor of k to the run time, making it n²k². Correct PAM is n²k, and FasterPAM is n² (all per iteration).

Assuming that you used k=10, a speedup of 10 should hence be easily possible by following the PAM algorithm more closely and computing the *change* in inertia directly, not the intertia. FasterPAM then removes redundant work in computing this change, because only one of the k centers can be (second) closest.

TimotheeMathieu commented:
@kno10 do you have a reference for PAM algorithm ? The ref I used was the original article (Clustering around Medoid) and the Wikipedia page. The Wikipedia page state that indeed, PAM's complexity is O(n^2k^2), which is the algorithm I followed, your version of PAM seems to already be an improvement on the basic PAM algorithm.

Maybe we could keep the basic PAM for completness's sake and because it can still be used on small examples but also implement one of the faster pam algorithm to make it more usable?

kno10 commented:
Kaufman, L. and Rousseeuw, P.J. (2008). Partitioning Around Medoids (Program PAM). In Finding Groups in Data (eds L. Kaufman and P.J. Rousseeuw). doi:10.1002/9780470316801.ch2
C.f. pages 102-103.

Wikipedia states (emphasis added):
> The runtime complexity of the original PAM algorithm per iteration of (3) is **O(k(n-k)²)**, by only computing the change in cost. A **naive** implementation recomputing the entire cost function every time will be in **O(n²k²)**.

Recomputing inertia is the naive approach. Feel free to make this more explicit in the Wikipedia article.

rth commented:
> Maybe we could keep the basic PAM for completness's sake and because it can still be used on small examples but also implement one of the faster pam algorithm to make it more usable?

+1 to have both versions.

TimotheeMathieu commented:
@kno10 thank you for the reference. 
I implemented the pam algorithm in O(kn^2) from "Partitioning Around Medoids (Program PAM). In Finding Groups in Data " however it is not very fast on digits dataset. I don't really know why, theoretically it should be faster than the naive implementation maybe I made a mistake in my code.

kno10 commented:
Well, the code is all but optimized. Lots of unnecessary data structures, copies, masking, redundancies in the computation of the cases, etc.

The name "fast" for the k-means-style approach is also a bad idea, because it is not FastPAM. Its known in classic FLP literature als "alternate".

Oh, and don't forget **initialization**! Your implementation chooses a very poor starting point, so that will also take many additional swaps to optimize - several times longer. You can roughly estimate there are about k iterations necessary from a bad starting condition to achieve the quality of the BUILD initialization strategy. BUILD is also an integral part of PAM.

TimotheeMathieu commented:
Ok. I give up this PR.

kno10 commented:
No, don't! I am not a contributor here (as I use python primarily for scripting, not for performance oriented code), and I do think this project needs to eventually have the PAM algorithm.
I did not mean to be rude with "all but optimized", but I'll try to flag some lines above to give you more precise critique.

TimotheeMathieu commented:
Ok, but please keep in mind that contributor work for free, they do not all code for a living, please try to use a less aggressive tone. 
I also do not do performance oriented code but I try to learn.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -190,8 +196,25 @@ def fit(self, X, y=None):
+                    np.arange(len(D) - self.n_clusters),
+                )
+                # transform generator to list
+                couples = [(i, j) for i, j in couples]
```
This creates (N-k)*k tuple objects, each storing two integer objects (the tuple totaling to about 128 bytes each + 8 for the reference in the list), each iteration. So this causes a substantial amount of both memory allocation and garbage collection. For MNIST (60000*10*136) this is about 86 MB storing trivial integer combinations. This is unnecessary, as only a single tuple is of interest, the minimum.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -112,12 +116,14 @@ def __init__(
         self,
         n_clusters=8,
         metric="euclidean",
+        method="fast",
```
Use "alternating", as this might be confused with FastPAM otherwise.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+        """Compute cost change for all the possible swaps"""
+        not_medoid_idxs = list(set(np.arange(len(D))) - set(medoid_idxs))
+
+        delta_costs = []
```
Don't use an array, store the optimum only.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
 
+    def _compute_optimal_swap(self, D, medoid_idxs, couples):
+        """Compute cost change for all the possible swaps"""
+        not_medoid_idxs = list(set(np.arange(len(D))) - set(medoid_idxs))
```
O(N*k) operation allocating O(N) memory that is not necessary.

TimotheeMathieu replied:
I am not sure what you mean by this. I did some changes to the code so maybe it is resolved but if not, could you explain more please ?

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+        delta_costs = []
+
+        # Compute the distance to the first and second closest medoid.
+        Djs, Ejs = np.sort(D[medoid_idxs], axis=0)[[0, 1]]
```
Recomputed each time with O(N k log(k)) time and O(Nk) memory. Cache this information to save this cost.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+        Djs, Ejs = np.sort(D[medoid_idxs], axis=0)[[0, 1]]
+
+        # Compute the change in cost for each swap.
+        for i, h in couples:
```
Just use two for loops instead of materializing this.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+                D[id_js, id_h] >= Ejs[id_js]
+            )
+            case3 = (D[id_i, id_js] > Djs[id_js]) & (
+                D[id_js, id_h] == Djs[id_js]
```
The `Djs` is likely supposed to be `Ejs`?

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            case1 = (D[id_i, id_js] == Djs[id_js]) & (
+                D[id_js, id_h] < Ejs[id_js]
+            )
+            case2 = (D[id_i, id_js] == Djs[id_js]) & (
```
Recomputing `D[id_i, id_js] == Djs[id_js]` a second time.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            case2 = (D[id_i, id_js] == Djs[id_js]) & (
+                D[id_js, id_h] >= Ejs[id_js]
+            )
+            case3 = (D[id_i, id_js] > Djs[id_js]) & (
```
Almost the same as `D[id_i, id_js] == Djs[id_js]` above, third time.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            case3 = (D[id_i, id_js] > Djs[id_js]) & (
+                D[id_js, id_h] == Djs[id_js]
+            )
+            C[case1] = (D[id_js, id_h] - Djs[id_js])[case1]
```
Note that computing these differences is almost the same as comparing them (and then if >0), so this again may have redundancies to above operations. But given the cost of interpreted python, I'd rather write these with `if` statements in Cython; pure Python will likely be too slow, and this form of vectorizing this with numpy may well be the second best choice.

TimotheeMathieu replied:
Maybe later someone will propose some cython amelioration, for now I do not know cython so I will not try this. I keep it the same.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            C[case1] = (D[id_js, id_h] - Djs[id_js])[case1]
+            C[case2] = (Ejs[id_js] - Djs[id_js])[case2]
+            C[case3] = (D[id_h, id_js] - Djs[id_js])[case3]
+            T = np.sum(C)
```
Since we only need the sum anyway, why store the result in `C` at all? Just sum each case individually, and them sum the three cases?

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+
+            delta_costs.append(T)
+        # If one of the swap decrease the objective, return that swap.
+        if np.min(delta_costs) < 0:
```
finding the minimum once, O((N-k)*k) cost.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            delta_costs.append(T)
+        # If one of the swap decrease the objective, return that swap.
+        if np.min(delta_costs) < 0:
+            i, h = couples[np.argmin(delta_costs)]
```
Finding the same minimum again, O((N-k)*k) cost again.

rth commented:
Yes, the wording in the comment was not ideal, but I'm glad that it was resolved :)

@TimotheeMathieu I am very grateful for the work you have been doing lately on this library!

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +275,50 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            C[case3] = (D[id_h, id_js] - Djs[id_js])[case3]
+            T = np.sum(C)
+
+            delta_costs.append(T)
```
Only store the minimum, don't build another `O((N-k)*k)` list. Here, `delta_cost` is a python list (not numpy), storing O((N-k)*k)) values, each using about 24+8 bytes. Using a numpy list would reduce this by a factor of 4, preallocation would avoid resizing the memory several times. But in the end, only the minimum is of interest.

TimotheeMathieu commented:
Thanks for the review @kno10 , I learned some tricks for faster python programming and corrected some mistakes. 
The algorithm became indeed faster. It is still slower than `alternating` method which is normal, maybe further improvements could be made using cython. 

On the `plot_kmedoids_digits.py` example:
`pam`: 385.29 seconds
`alternating`: 3.16 seconds

This is much better than the first implementation but still slow. This is not necessarily surprising because the complexity is indeed theoretically greater for `pam` than for `alternating`. 

Also, the computation of the optimal swap is not done in parallel even though it would be easy to do with joblib (which is already part of the requirements for scikit-learn). Is it interesting to do it or is it not worth it because when we implement in cython it would do it automatically (I know nothing of cython, maybe it is a naive question).

@rth could you give me the code you used for the benchmarking you did in the third post of this PR please ? It is very nice, I would like to use it here and I would also like to use it to troubleshoot the robust module (to improve PR #77 ).

TimotheeMathieu commented:
Rectification: there is a problem, the algorithm is two times slower than the naive implementation I first did so there seems to be a problem in the code.

rth commented:
> could you give me the code you used for the benchmarking you did in the third post of this PR please ?

I used cProfile + [snakeviz](https://jiffyclub.github.io/snakeviz/) to do that profiling:
```
python -m cProfile -o out.prof example_script.py
snakeviz out.prof
```
which starts a web app to visualize the profiling results.


> I obtain more or less the same plot for the three algorithms: alternating, naive pam and pam from last commit so the algorithm gives a good result. Just the computational time is weird.

It's reassuring that results are consistent. I can help you port the code to Cython letter, if needed, but let's see if we can still improve the performancec it a bit in pure Python & numpy. 

It would be good to add a test that checks that on a given (small) dataset the same clusters are found with different methods.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -190,8 +195,27 @@ def fit(self, X, y=None):
+                )
+
+                # Compute the distance to the first and second closest medoid.
+                Djs, Ejs = np.sort(D[medoid_idxs], axis=0)[[0, 1]]
```
This will run every iteration again. Do this only once, then update the values.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -190,8 +195,27 @@ def fit(self, X, y=None):
+                # Update medoids with the new cluster indices
+                self._update_medoid_idxs_in_place(D, labels, medoid_idxs)
+            elif self.method == "pam":
+                not_medoid_idxs = list(
```
Instead of materializing this list each iteration, I would simply check for medoids in the actual loop.
If I recall correctly, the original PAM fortran code used an array of size N to indicate which objects are medoids.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +276,48 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+            for h in range(len(D) - self.n_clusters):
+                id_i = medoid_idxs[i]
+                id_h = not_medoid_idxs[h]
+                id_js = np.delete(not_medoid_idxs, h)
```
This will make a copy, and run in O(n)

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -252,6 +276,48 @@ def _update_medoid_idxs_in_place(self, D, labels, medoid_idxs):
+                case2 = cluster_i_bool & ~second_best_medoid
+                case3 = ~cluster_i_bool & (D[id_js, id_h] < Djs[id_js])
+
+                C1 = np.sum((D[id_js, id_h] - Djs[id_js])[case1])
```
It would be better to do this without the double indexing on id_js and then case1. A single index should be more performant.

TimotheeMathieu commented:
I think I am near the end.

I learned some cython and I implemented pam in cython. After some work, I obtained performances similar to R algorithm pam from library cluster (that I consider state of the art for pam) and at least this gave a smaller time than naive method when n_clusters is small !

I also implemented parallel computing in cython, now there is a n_threads argument to K_Medoid that is default to all the CPU available (it may be better to default to 1 core ?) as a result this algorithm is even faster than R algorithm ! Although there is a problem I don't understand, the checks seems to indicate that openmp does not work on windows and mac, I don't know if it is a known problem for now I just stated in the setup.py to enable openmp for linux only.

Result on digits database example (I do the fit for the three metrics): 
```
             wall_time
method                
pam           5.053187
alternating   0.097380

```
which is approximately a 200x speedup compared to previous pam implementation.

@rth : concerning the tests, in fact alternating and pam give different clusters. The centers are rather close to one another but they are not the same. I devised a test that checked that the distance between the centers for the two methods are not far away from one another on a sub-sample of the digits dataset but this test fails in the CI when using windows.

kno10 commented on sklearn_extra/cluster/_k_medoids_swap.pyx:
```diff
@@ -0,0 +1,68 @@
+            T = 0
+            # compute for all not-selected points the change in cost
+            for j in prange(not_medoid_shape, nogil=True, num_threads = n_threads):
+                to_add = 0
```
remove this variable. Just add to T?

TimotheeMathieu replied:
I can't because this loop is with nogil, it may seems like a hack but this allowed me to make the loop able to compute in parallel while if I use T directly the nogil fails and this is at practically no cost so I kept it.

kno10 replied:
I don't see why this makes a difference: either all three increments can be made atomic, or the one that you have right now cannot be made atomic either.
As for parallelization with *threads*, I would rather do this on the outermost loop. The parallel inner loop will be called (N-k)k times, and do comparatively little work then each time. The (N-k) outer loop should be fine-grained enough for thread level parallelism.

TimotheeMathieu replied:
Ok, done.

kno10 commented on sklearn_extra/cluster/_k_medoids_swap.pyx:
```diff
@@ -0,0 +1,68 @@
+
+    # If one of the swap decrease the objective, return that swap.
+    if best_cost_change < 0:
+        return best_couple
```
I would also return the cost change, make best_couple a triple.

TimotheeMathieu replied:
why ? I don't have the initial cost so this change in cost is not very useful for me outside this function.

kno10 replied:
You can easily get the initial cost when doing the first assignment. The cost change will give you the cost after the swap (although this might accumulate some error, I guess). Then you also get the final cost "for free".
With k-means people like using the cost reduction for early stopping (although I wouldn't recommend this here, as the cost decrease might not be very smooth). So I wouldn't be surprised if you get a request for a `tol` parameter at some point.
Last but not least, there could be some odd corner case where because of floating point numerics the cost change with two equivalent positions ends up not being exactly zero, but -1e-100 or something similarly small, I guess. I do not have a test case where this actually happens though.

For unit testing, the cost value is also worth testing. IIRC you mentioned you saw platform differences - do they also affect quality?

TimotheeMathieu replied:
Ok done.

The difference between platforms is only for the parallel loop, a problem to compile the library needed. I looked into it and scikit-learn setup.py seems to have something that solves it but I don't understand how they do it and I don't cant to copy/paste things that are not needed so for now I keep it parallel for linux only.

kno10 commented on sklearn_extra/cluster/_k_medoids.py:
```diff
@@ -199,16 +212,32 @@ def fit(self, X, y=None):
-                # Compute the distance to the first and second closest medoid.
-                Djs, Ejs = np.sort(D[medoid_idxs], axis=0)[[0, 1]]
+                    # update Djs and Ejs with new medoids
+                    Djs, Ejs = np.sort(D[medoid_idxs], axis=0)[[0, 1]]
```
This update is O(Nk log k). It can be reduced to O(Nk) by using numpy.partition (but the resulting code will not be pretty).
But as many points will not be affected by the swap, this can be further optimized the very same way as finding the swap in PAM, to yield an algorithm that for most points can avoid iterating over all medoids.
But this part will likely need to be written in cython then to be fast.

TimotheeMathieu replied:
Here is a profiling of the function now. 
![image](https://user-images.githubusercontent.com/30346931/99142621-0ef6d600-2657-11eb-8d69-0bc9cdf0a5af.png)

the part with this computation is the rectangle on the second line, third column, it takes practically no time !
You already commented on Djs and Ejs but as you see in the profiling it is not worth the work. In kmedoid now, we have 0.35% of the time dedicated to other things than `compute_optimal_swap` and `validation`, I don't think this needs further improvement.

TimotheeMathieu commented:
@rth and @kno10 are you all right with the implementation now ? 
Can I merge it or do you want to work a little more on this before the merge ?

I think a future work would be to implement faster pam algorithm but it may not be the most pressing matter for now because we already have a fast kmedoid implementation with the alternating method.

adrinjalali commented:
Generally we try to avoid merging our own PRs. So I wouldn't do that :)

TimotheeMathieu commented:
Ok, I didn't know that, thank you for the information.

rth commented:
Yes, sorry I should have been more explicit in my email. We still need at least one approved review to merge PR. This is particularly important for major changes (new algorithm, new solvers, anything that changes the public API or makes significant code changes). For small documentation improvements, and CI fixes, they can be merged without review after a couple of days if you see that no one objects.

It's a balance between ensuring the merged code and algorithms are sound, and given the few people who review code in this repo and it being more experimental than scikit-learn moving forward with development.

In any case, great work on this PR @TimotheeMathieu! I'll try to review in the next few days.

kno10 commented:
I think the pull request is definitely a major improvement and addition to the package. Given the significance of PAM, I believe it should eventually go to the main sklearn branch though once BUILD initialization is added, too. PAM certainly qualifies then.

For testing, I suggest to add some of the well-known harder problems and also evaluate quality.

In particular, I suggest using this data set:
http://people.brunel.ac.uk/~mastjjb/jeb/orlib/files/pmed40.txt

PAM with BUILD usually (depends on the data permutation; because of tied distances in this data set, order matters) found a solution around 5140, the best of 11 runs was 5139. The known optimum is 5128. K-means-style iterations with random initialization was usually around 6233, with the best of 11 runs 5947, because of the known limitations. Hence this is a data set that can show clear quality differences between the methods. BUILD alone already scores around 5189, giving much better results than k-means-style alternating, too.

TimotheeMathieu commented:
Ok, I tested on the problem suggested by @kno10 .
Results on one run:
```
pam  
        random: 7158
        heuristic : 7825
        k-medoids++: 6226

alternating: 
                   random: 6392
                   k-medoid++: 5627
                   heuristic: 6661
```

From what I see, for now we are far from optimal on this problem. 
Do I implement BUILD in this PR or do I begin another one ? I think we can continue in this one because, as @kno10 said, BUILD is part of PAM so for now I didn't implement the whole PAM algorithm.

TimotheeMathieu commented:
So there was a bug with the parallel computing, I removed it. Maybe for later.

For now, new very good results :
```
pam  
        random: 7326 (maximum number of iterations attained, I didn't try to increase max_iter)
        heuristic : 5149
        k-medoids++: 5155
        build: 5139

alternating: 
                   random: 6214
                   k-medoid++: 5660
                   heuristic: 6661
                   build: 5176
```
running time
```
pam  
        random: 65042ms
        heuristic : 15167ms
        k-medoids++: 14091ms
        build: 3970ms

alternating: 
                   random: 13ms
                   k-medoid++: 17ms
                   heuristic: 11ms
                   build: 118ms

```

So (phew) the algo is good. Moreover, personally I am proud of myself because it is faster than R implementation for which, on my computer, pam+build run in 5875ms.

rth reviewed (commented):
Great work @TimotheeMathieu : 

Yes, let's for not now enable openmp, it's harder to distribute wheels with openmp. We can look into later.

In this PR we would need to update https://scikit-learn-extra.readthedocs.io/en/latest/user_guide.html#k-medoids which currently says that PAM is not implemented. Also add references there for the build method and maybe some explanations on the differences between different methods.

And finally it would be great to have some toy example as a unit-test that that checks that all 3 algorithms produce identical clustering (or that the cost function is close enough). 

Another few minor comments are below.

rth commented:
BTW @kno10 if you are interested in improving the KMedoids documentation (in a separate PR) that would also be very welcome.

kno10 commented:
Maybe make "build" the default, because of quality - "heuristic" does not fare very well, does it? If it was designed close to the Park and Jun approach, I'd suggest to name it this way. "heuristic" is a very unspecific name.
For completeness, you may also want to allow a "build only" use (as suggested in the PAM paper); the alternating approach may improve the quality only very little over build.

TimotheeMathieu commented:
I am not necessarily convinced that 'build' should be the default initialization because it is very non-robust : if there is an outlier, build will choose this outlier as a medoid and the algorithm would then have some difficulties to escape this outliers. In my opinion, one of the big advantages of k-medoids is its robustness and with build it is not robust (I added this comment in the doctring  of `KMedoids` if people want to use KMedoids with a dataset corrupted by outliers).

rth commented on sklearn_extra/cluster/tests/test_k_medoids.py:
```diff
@@ -12,10 +12,44 @@
+rng = np.random.RandomState(seed)
+X_cc, y_cc = make_blobs(
+    n_samples=100,
+    centers=np.array([[-1, -1], [1, 1]]),
```
Maybe we can reduce the `cluster_std` to 0.2 some such value? Currently it's at 1.0 and that's why there is an overlap between clusters and we detect some fraction of points in the wrong cluster. Or alternatively comment that the 0.8 value in test is due to the data and not the the kmedoids itself.

TimotheeMathieu replied:
I commented to explain. I prefer to keep the clusters not so separable in order to test the stability of the method at the same time.

kno10 commented:
> I am not necessarily convinced that 'build' should be the default initialization because it is very non-robust : if there is an outlier, build will choose this outlier as a medoid and the algorithm would then have some difficulties to escape this outliers. In my opinion, one of the big advantages of k-medoids is its robustness and with build it is not robust (I added this comment in the doctring of `KMedoids` if people want to use KMedoids with a dataset corrupted by outliers).

I understand that an "outlier cluster" may be "undesirable" from a user intuition.
But do you have an example where it actually is the *worse* solution, by the *objective* function?
I doubt that choosing worse starting conditions satisfies some statistical notion of "robustness".

From my understanding and experiments, the initialization should have rather little effect on the average result quality because of the way SWAP operates. Essentially, the initial solution (or a very similar solution) chosen by BUILD would often be reached by SWAP from many random initialization. BUILD just speeds this up, reducing the runtime usually by around k-few iterations.

The reason why FasterPAM does not use BUILD anymore is because its much more expensive than the improved swapping. In the experiments random initialization then was just as good (with the additional benefit of being randomized).

TimotheeMathieu commented:
Indeed, the solution is not worse by the objective function... when you compute it on the dataset with the outlier. 

> But do you have an example where it actually is the worse solution, by the objective function?

The idea in robust statistics is that outliers should not count so we would not compute the objective function on the outlier and in this case, yes it is easy to find examples.

> I doubt that choosing worse starting conditions satisfies some statistical notion of "robustness".

There are several notions of statistical robustness, let me present one of the most widespread informally. Let your data come from a mixture of three gaussians: two gaussians are informative data and one is outlier data (see the figure). The outlier gaussian blob contains only a small number of points. The idea is a sort of continuity of the estimator, i.e. a small change in the law of the data (here for instance with respect to the total variation distance between distribution) should not have a big impact on the result of the estimation. Here this means the estimation should return something similar to what the estimation is when there are no outliers.
For more references on this subject there are numerous books but a very well known is Robust Statistics: The Approach Based on Influence Functions by Frank R. Hampel, Elvezio M. Ronchetti, Peter J. Rousseeuw, Werner A. Stahel.

Here is a picture of what happen in practice with KMedoids PAM (notice that swap do not succeed in recovering the two inlier gaussian blobs) there are 97 inliers and 3 outliers. The color are the clusters found by the algorithm which is not what I wanted, I wanted to recover the two gaussian blobs and ignore the outliers (here outliers are in yellow). In practice, for instance you know that you are searching for two clusters but there are outliers caused by captor error, human error... 
![image](https://user-images.githubusercontent.com/30346931/99902563-fe7dd580-2cbe-11eb-89f1-aaa6af738eff.png)

rth commented:
OK, so since the documentation will be updated in https://github.com/scikit-learn-contrib/scikit-learn-extra/pull/81 this PR LGTM. 

The only remaining comment I have is https://github.com/scikit-learn-contrib/scikit-learn-extra/pull/73#discussion_r528311924

rth reviewed (approved):
Thanks @TimotheeMathieu

rth commented:
There is a failure in one of the CI jobs on master,

```
method = 'alternate', init = 'k-medoids++'

    def test_kmedoid_results(method, init):
        expected = np.hstack([np.zeros(50), np.ones(50)])
        km = KMedoids(n_clusters=2, init=init, method=method)
        km.fit(X_cc)
        # This test use data that are not perfectly separable so the
        # accuracy is not 1. Accuracy around 0.85
>       assert (np.mean(km.labels_ == expected) > 0.8) or (
            1 - np.mean(km.labels_ == expected) > 0.8
        )
E       assert (0.53 > 0.8 or (1 - 0.53) > 0.8)
```

which is consistent with known issues of the alternate method. We need to fix the random_state in the test.

kasparg commented:
:clap: Just commenting here that this PR is great improvement and a lifesaver!

kno10 commented:
I have published a Python wrapper for a Rust port of PAM and FasterPAM:
https://pypi.org/project/kmedoids/

On 20newsgroups_vectorized, cosine distance, k=20, Google Colab:
| Package | Algorithm | Runtime [ms] | Loss |
|---|---|---:|---:|
|sklearn | cosine | 13262.62 | |
| scikit-learn-extra 0.1.0b2 | Alternating | 2427.86 | 5194.06 |
| scikit-learn-extra 0.1.0b2 | Alt. Heurstic | 668.57 | 5137.16 |
| kmedoids 0.1.5 | Alternating | 1705.96 | 5194.06 |
| kmedoids 0.1.5 | FasterPAM | 344.97 | 4996.85 |
| kmedoids 0.1.5 | PAM | 33883.32 | 4994.73 |
| kmedoids 0.1.5 | BUILD | 4640.23 | 4995.06 |
| kmedoids 0.1.5 | random init | 6.14 | 5751.77 |

Averages of 20 iterations. PAM and BUILD are deterministic; all others are random seeded (except "heuristic").
On other data sets, PAM was worse than FasterPAM; this is a give or take based on starting conditions and local optima. Best of FasterPAM with 20 restarts was 4994.81; but I intend to rerun this benchmark with shuffled data as there may well be some order dependency on the 20news data. This will be worth studying, if the data order affects which local optima can be found with random initialization.

I am a bit surprised how much runtime difference the "heuristic" initialization makes here; yet the quality is still considerably worse than the PAM algorithms. In the FasterPAM experiments, initialization had little effect on quality, so choosing the fastest (random) was reasonable. And you can see that BUILD alone finds better results than Alternating.
